hkmc2/shared/src/main/scala/hkmc2/codegen/Block.scala

```diff
@@ -51,15 +53,142 @@ sealed abstract class Block extends Product with AutoLocated:
+    case HandleBlockReturn(res) => f(res).map(_(HandleBlockReturn(_))).getOrElse(this)
+    case _: End | _: Break | _: Continue | _: Match | _: Label | _: Begin | _: TryBlock | _: Define | _: HandleBlock | _: HandleBlockReturn => this
+  
+  def mapPath(f: Path => Path): Block = this match
+    case Return(res: Path, implct) => Return(f(res), implct)
+    case Throw(exc: Path) => Throw(f(exc))
+    case Assign(l, r: Path, rst) => Assign(l, f(r), rst)
+    case b @ AssignField(l, n, r: Path, rst) => AssignField(l, n, f(r), rst)(b.symbol)
+    case Match(scrut: Path, arms, dflt, rst) => Match(f(scrut), arms, dflt, rst)
+    case Define(ValDefn(owner, k, sym, rhs: Path), rst) => Define(ValDefn(owner, k, sym, f(rhs)), rst)
+    case HandleBlockReturn(res: Path) => HandleBlockReturn(f(res))
+    case _: Return | _: Throw | _: Assign | _: AssignField | _: End | _: Break | _: Continue | _: Match | _: Label | _: Begin | _: TryBlock | _: Define | _: HandleBlock | _: HandleBlockReturn => this.mapResult {
+      case r @ Call(fun, args) => S(_(Call(f(fun), args.map(arg => Arg(arg.spread, f(arg.value))))(r.isMlsFun)))
+      case r @ Instantiate(cls, args) => S(_(Instantiate(f(cls), args.map(f))))
+      case r => N
+    }
+  
+  def mapSyms(using State, SymbolSubst): Block =
+    def pMap(p: Param) = 
+      val newSym: LocalSymbol & NamedSymbol = p.sym.subst
+      Param(p.flags, p.sym.subst, p.sign)
+    def pListMap(p: ParamList) = ParamList(p.flags, p.params.map(pMap), p.restParam.map(pMap))
+    def pathMap(p: Path): Path = p match
+      case s @ Select(qual, name) => Select(pathMap(qual), name)(s.symbol.map(_.subst))
+      case Value.Ref(l) => Value.Ref(l.subst)
+      case Value.This(sym) => Value.This(sym.subst)
+      case Value.Lit(lit) => p
+      case Value.Lam(params, body) => Value.Lam(
+        pListMap(params),
+        body.mapSyms
+      )
+      case Value.Arr(elems) => Value.Arr(elems.map(a => Arg(a.spread, pathMap(a.value))))
+
+    def _resMap(r: Result) = r match
+      case p: Path => pathMap(p)
+      case c @ Call(fun, args) => Call(pathMap(fun), args.map(a => Arg(a.spread, pathMap(a.value))))(c.isMlsFun)
+      case Instantiate(cls, args) => Instantiate(pathMap(cls), args.map(pathMap))
+
+    def resMap[T <: Result : ClassTag](l: T): T = _resMap(l) match
+      case r : T => r
+      case _ => l
+
+    this match
+      case Return(res, implct) => Return(resMap(res), implct)
+      case Throw(exc: Path) => Throw(pathMap(exc))
+      case Assign(l, r, rst) => Assign(l.subst, resMap(r), rst.mapSyms)
+      case blk @ AssignField(l, n, r, rst) => AssignField(pathMap(l), n, _resMap(r), rst.mapSyms)(blk.symbol)
+      case Match(scrut, arms, dflt, rst) => 
+        val newArms = arms.map((cse, blk) =>
+          val newCse = cse match
+            case Case.Lit(lit) => cse
+            // HACK: cls.subst gives something other than ClassSymbol | ModuleSymbol and it doesn't type check...
```

Comment:
Another limitation of Scala, though it is not too severe. I wouldn't call this a hack, just a mostly benign workaround. This could be fixed by introducing yet another parent trait and using that instead of `ClassSymbol | ModuleSymbol`.